src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -4063,9 +4063,15 @@ LogicalResult ONNXIsInfOp::inferShapes(
   return emitError(NOT_IMPLEMENTED_MESSAGE);
 }
 
+//===------------------------------------------------------------------------===//
+// IsNaNOp
+//===------------------------------------------------------------------------===//
 LogicalResult ONNXIsNaNOp::inferShapes(
     std::function<void(mlir::Region &)> doShapeInference) {
-  return emitError(NOT_IMPLEMENTED_MESSAGE);
+  auto inputShape = X().getType().cast<ShapedType>().getShape();
```

Comment:
I would add the following lines
```cpp
    if (!hasShapeAndRank(X()))
      return success();
```
Otherwise when the input to `IsNaN` is unranked, the shape inference will be incorrect. Here is a test example that fails for this reason. (This can be run by putting it in the `onnx_shape_inference.mlir` file)
```MLIR
//===----------------------------------------------------------------------===//
/// Test shape inference for IsNaNOp.
//===----------------------------------------------------------------------===//
func @test_isnan1(%arg0 : tensor<*xf32>) -> tensor<*xi1> {
  %0 = "onnx.IsNaN"(%arg0) : (tensor<*xf32>) -> tensor<*xi1>
  return %0 : tensor<*xi1>

  // CHECK-LABEL: test_isnan1
  // CHECK: [[RES:%.+]] = "onnx.IsNaN"(%arg0) : (tensor<*xf32>) -> tensor<*xi1>
  // CHECK: return [[RES]] : tensor<*xi1>
}
```